runtime/sys.c

```diff
@@ -286,10 +288,34 @@ CAMLprim value caml_sys_remove(value name)
   CAMLparam1(name);
   char_os * p;
   int ret;
+
   caml_sys_check_path(name);
   p = caml_stat_strdup_to_os(String_val(name));
   caml_enter_blocking_section();
   ret = unlink_os(p);
+#ifdef _WIN32
+  /* On Windows, trying to unlink a symlink to a directory will return
+   * EACCES, but the symlink can be deleted with rmdir. */
+  if (ret != 0 && errno == EACCES) {
+    if ((ret = GetFileAttributes(p)) != INVALID_FILE_ATTRIBUTES) {
+      if ((ret & FILE_ATTRIBUTE_REPARSE_POINT) &&
+          (ret & FILE_ATTRIBUTE_DIRECTORY)) {
+        ret = rmdir_os(p);
+      } else if (ret & FILE_ATTRIBUTE_DIRECTORY) {
+        /* POSIX uses EPERM rather than EACCES when the path is a
+         * directory. */
+        errno = EPERM;
+        ret = -1;
+      } else {
+        errno = EACCES;
+        ret = -1;
+      }
+    } else {
+      errno = EACCES;
+      ret = -1;
+    }
+  }
+#endif
```

Comment:
This adds rather too much `_WIN32` to `sys.c`. `unlink_os` is sadly used in ocamlyacc, so I think the best course of action is to add a definition in `misc.h` for `caml_unlink` which will be `unlink` on Unix and `caml_win32_unlink` on Windows. You can then put the declaration of `caml_win32_unlink` in `osdeps.h` and move this function to `win32.c`.

If `otherlibs/unix/unlink.c` then uses `caml_unlink` instead of `unlink_os`, then the win32unix gets the change "for free" and I think all the changes in `otherlibs/win32unix/unix.ml` aren't needed.